### PR TITLE
HHH-15422 Pick up `CurrentTenantIdentifierResolver` and `MultiTenantConnectionProvider` from BeanContainer if not explicit set

### PR DESCRIPTION
**Pull request by quaff**
[Please describe here what your change is about]

----------------------
By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt)
and can be relicensed under the terms of the [LGPL v2.1 license](https://www.gnu.org/licenses/old-licenses/lgpl-2.1.txt) in the future at the maintainers' discretion.
For more information on licensing, please check [here](https://github.com/hibernate/hibernate-orm/blob/main/CONTRIBUTING.md#legal).

----------------------

https://hibernate.atlassian.net/browse/HHH-15422

### CODE REVIEW

**Comment by gavinking** [comment]
So we have not up until this point allowed `ConnectionProvider`s to be beans, and so:

1. I wonder why it makes sense to do this for the special case of `MultiTenantConnectionProvider` if we don't do it in general. What's special about `MultiTenantConnectionProvider`? I assume the reason is that you want to `@Inject` into it. But inject what precisely? Tell me a story here. Given me an example.
2. We have been having troubles with chicken/egg problems with respect to the CDI integration we already have, and so I'm sorta disinclined to add new functionality to that until we feel like we have this stuff licked.

I dunno, WDYT @sebersole @scottmarlow @yrodiere?

**Comment by yrodiere** [comment]
> I wonder why it makes sense to do this for the special case of `MultiTenantConnectionProvider` if we don't do it in general. What's special about `MultiTenantConnectionProvider`? I assume the reason is that you want to `@Inject` into it. But inject what precisely? Tell me a story here. Given me an example.

To others reading this: the conversation continued on the Jira, see https://hibernate.atlassian.net/browse/HHH-15422?focusedCommentId=116437

But IMO the main point here is that a `MultiTenantConnectionProvider` needs access to datasources, which are typically created by the framework, and thus are more easily retrieved through injection.

> We have been having troubles with chicken/egg problems with respect to the CDI integration we already have

That's a good point, too much CDI reliance can be problematic in Jakarta application servers, or at least in WildFly.

I would add that relying on CDI in very early phases of bootstrap is problematic for Quarkus as well, if we want to move more of the bootstrap to build time. Though this specific PR seems fine in that regard.

> and so I'm sorta disinclined to add new functionality to that until we feel like we have this stuff licked.

I kind of agree, but I don't think that's a healthy approach unless we have a clearly identified Jira issue and at least the beginning of a plan to solve that chicken/egg problem.
If we don't intend to solve that chicken/egg problem, then we probably need to record the criteria for further CDI integration somewhere (must be lazy, or must happen after a certain point in bootstrap, or ...).

File: hibernate-core/src/main/java/org/hibernate/boot/internal/SessionFactoryOptionsBuilder.java [hunk 1]
```diff
@@ -376,6 +381,38 @@ public SessionFactoryOptionsBuilder(StandardServiceRegistry serviceRegistry, Boo
 				CurrentTenantIdentifierResolver.class,
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/boot/internal/SessionFactoryOptionsBuilder.java:381]: This should IMO be resolved as bean as well i.e.

```java
		final BeanContainer beanContainer = serviceRegistry.requireService( ManagedBeanRegistry.class ).getBeanContainer();
		if ( beanContainer != null ) {
			this.currentTenantIdentifierResolver = strategySelector.resolveStrategy(
				CurrentTenantIdentifierResolver.class,
				configurationSettings.get( MULTI_TENANT_IDENTIFIER_RESOLVER ),
				null,
				clazz -> beanContainer.getBean( 
					clazz, 
					new BeanContainer.LifecycleOptions() {
						@Override
						public boolean canUseCachedReferences() {
							return true;
						}

						@Override
						public boolean useJpaCompliantCreation() {
							return false;
						}
					}, 
					FallbackBeanInstanceProducer.INSTANCE
				)
			);
		}
		else {
			this.currentTenantIdentifierResolver = strategySelector.resolveStrategy(
				CurrentTenantIdentifierResolver.class,
				configurationSettings.get( MULTI_TENANT_IDENTIFIER_RESOLVER )
			);
		}
```

File: hibernate-core/src/test/java/org/hibernate/orm/test/idgen/userdefined/UserDefinedGeneratorsTests.java [hunk 2]
```diff
@@ -61,10 +62,16 @@ public void testCreateGeneratorsByBeanContainer() {
 		
 		final BeanContainer beanContainer = Mockito.mock( BeanContainer.class );
 		given(beanContainer.getBean( any(), any(), any() ) ).willAnswer( invocation -> {
+			Class<?> beanType = (Class<?>) invocation.getArguments()[0];
 			LifecycleOptions options = (LifecycleOptions) invocation.getArguments()[1];
-			assertThat( options.canUseCachedReferences(), is( false ) );
-			assertThat( options.useJpaCompliantCreation(), is( true ) );
-			return (ContainedBean<?>) TestIdentifierGenerator::new;
+			if (beanType == TestIdentifierGenerator.class) {
```
Comment (beikov) [inline hibernate-core/src/test/java/org/hibernate/orm/test/idgen/userdefined/UserDefinedGeneratorsTests.java:67]: Why are these changes needed?

File: hibernate-core/src/test/java/org/hibernate/orm/test/idgen/userdefined/SimpleBeanContainer.java [hunk 3]
```diff
@@ -19,10 +21,8 @@ public <B> ContainedBean<B> getBean(
 			Class<B> beanType,
 			LifecycleOptions lifecycleOptions,
 			BeanInstanceProducer fallbackProducer) {
-		if ( beanType == SimpleGenerator.class ) {
-			return () -> (B) new SimpleGenerator( new AtomicLong( INITIAL_VALUE ) );
-		}
-		return null;
+		return () -> (B) ( beanType == SimpleGenerator.class ?
```
Comment (beikov) [inline hibernate-core/src/test/java/org/hibernate/orm/test/idgen/userdefined/SimpleBeanContainer.java:24]: Why are these changes needed?

File: hibernate-core/src/main/java/org/hibernate/engine/jdbc/connections/internal/MultiTenantConnectionProviderInitiator.java [hunk 4]
```diff
@@ -41,7 +45,36 @@ public Class<MultiTenantConnectionProvider<?>> getServiceInitiated() {
 	@Override
 	public MultiTenantConnectionProvider<?> initiateService(Map<String, Object> configurationValues, ServiceRegistryImplementor registry) {
 		if ( !configurationValues.containsKey( AvailableSettings.MULTI_TENANT_CONNECTION_PROVIDER ) ) {
-			// nothing to do, but given the separate hierarchies have to handle this here.
+			final BeanContainer beanContainer = Helper.allowExtensionsInCdi( registry ) ? registry.requireService( ManagedBeanRegistry.class ).getBeanContainer() : null;
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/engine/jdbc/connections/internal/MultiTenantConnectionProviderInitiator.java:48]: I think a similar change should be done for `ConnectionProviderInitiator`

File: hibernate-core/src/main/java/org/hibernate/engine/jdbc/connections/internal/MultiTenantConnectionProviderInitiator.java
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/engine/jdbc/connections/internal/MultiTenantConnectionProviderInitiator.java:1]: Also, I think that `MultiTenantConnectionProvider` and `ConnectionProvider` should be instantiated through the bean container if possible i.e.

This should IMO be resolved as bean as well i.e.

```java
	final BeanContainer beanContainer = serviceRegistry.requireService( ManagedBeanRegistry.class ).getBeanContainer();
	if ( beanContainer != null ) {
		return beanContainer.getBean( 
			implClass, 
			new BeanContainer.LifecycleOptions() {
				@Override
				public boolean canUseCachedReferences() {
					return true;
				}

				@Override
				public boolean useJpaCompliantCreation() {
					return false;
				}
			}, 
			FallbackBeanInstanceProducer.INSTANCE
		);
	}
	else {
		try {
			return implClass.newInstance();
		}
		catch (Exception e) {
			log.warn( "Unable to instantiate specified class [" + implClass.getName() + "]", e );
			throw new ServiceException( "Unable to instantiate specified multi-tenant connection provider [" + implClass.getName() + "]" );
		}
	}
```

**Review by gavinking — CHANGES_REQUESTED** [review]
Please redo this to use the mechanism described here:

https://jakarta.ee/specifications/persistence/3.2/jakarta-persistence-spec-3.2#entity-listeners

and here:

https://jakarta.ee/specifications/persistence/3.2/jakarta-persistence-spec-3.2#a2999

These passages of the JPA spec describe how to enable injection into a persistence-managed object _without_ having to make that object into a CDI bean.

This is actually what we should always be doing in this sort of situation.

**Review by beikov — CHANGES_REQUESTED** [review]

File: hibernate-core/src/main/java/org/hibernate/engine/jdbc/connections/internal/MultiTenantConnectionProviderInitiator.java
Comment (gavinking) [reply hibernate-core/src/main/java/org/hibernate/engine/jdbc/connections/internal/MultiTenantConnectionProviderInitiator.java:1]: No, none of these objects should be beans. See my review comment.

Things do not need to be beans in order to have beans injected into them. The JPA spec goes to some lengths to make that clear, and we should not be ignoring it and just doing our own thing.

**Comment by gavinking** [comment]
To be clear, because there's apparently a lot of historical confusion on this point.

There's three strong reasons why something might need to be a bean:

1. it's going to be injected into other things,
2. it's contextual, i.e., it has a scope, or
3. it needs to have interceptors to handle transaction management or whatever.

Neither of those motivations apply to the objects we're talking about here, and so they do not need to be beans.

File: hibernate-core/src/main/java/org/hibernate/engine/jdbc/connections/internal/MultiTenantConnectionProviderInitiator.java
Comment (beikov) [reply hibernate-core/src/main/java/org/hibernate/engine/jdbc/connections/internal/MultiTenantConnectionProviderInitiator.java:1]: I agree that we should also support the case of injection into a non-bean, but we should IMO still try to instantiate classes as beans if a bean container is available.

Comment (gavinking) [reply hibernate-core/src/main/java/org/hibernate/engine/jdbc/connections/internal/MultiTenantConnectionProviderInitiator.java:1]: No we should not, for the exact same reasons that the JPA spec says that attribute converters and entity listeners are not instantiated as beans. The JPA group had very good reasons to set things up this way, as we have seen.

**Comment by beikov** [comment]
If you think about a `MultiTenantConnectionProvider` which can delegate it's work to multiple datasource, it is highly desirable to make it a bean to inject datasources into it.
I just don't see a reason why we wouldn't want to make these things beans if we clearly know they could be.

**Comment by gavinking** [comment]
> If you think about a `MultiTenantConnectionProvider` which can delegate it's work to multiple datasource, it is highly desirable to make it a bean to inject datasources into it.

I'm not sure if you've simply not read my comments, or if you've not understood them. But the whole point is that:

> Things do not need to be beans in order to have beans injected into them. 

Please go and read those sections of the JPA spec which explain how to do it. (We even already have an implementation of that in Hibernate, by the way.)


> I just don't see a reason why we wouldn't want to make these things beans if we clearly know they could be.

Because that ties the lifecycle of Hibernate with the lifecycle of CDI and creates problems.

**Comment by gavinking** [comment]
Note that my objections to this pull request actually boil down to only the following, which occurs in two places:

```java
@Override
public boolean useJpaCompliantCreation() {
	return false;
}
```

If we change those to `return true;`, I think my objections are addressed, and I'm fine with doing this.

File: hibernate-core/src/test/java/org/hibernate/orm/test/idgen/userdefined/SimpleBeanContainer.java (hunk 3 above)
Comment (quaff) [reply hibernate-core/src/test/java/org/hibernate/orm/test/idgen/userdefined/SimpleBeanContainer.java:24]: It will fail new tests, I was wrong, It shouldn't return `null`.

File: hibernate-core/src/test/java/org/hibernate/orm/test/idgen/userdefined/UserDefinedGeneratorsTests.java (hunk 2 above)
Comment (quaff) [reply hibernate-core/src/test/java/org/hibernate/orm/test/idgen/userdefined/UserDefinedGeneratorsTests.java:67]: It will fail new tests.

Comment (quaff) [reply hibernate-core/src/test/java/org/hibernate/orm/test/idgen/userdefined/UserDefinedGeneratorsTests.java:67]: It fails because of the changes, this test was not robust enough.

**Comment by quaff** [comment]
> So it turns out that my objections to this pull request actually boil down to only the following, which occurs in two places:
> 
> ```java
> @Override
> public boolean useJpaCompliantCreation() {
> 	return false;
> }
> ```
> 
> If we change those to `return true;`, I think my objections are addressed, and I'm fine with doing this.
> 
> (Note that this is consistent with every other implementation of `BeanContainer.LifecycleOptions` internal to Hibernate.)

Quoted from `JpaCompliantLifecycleStrategy.java`
```
 * In particular, {@code @Singleton}-scoped or {@code @ApplicationScoped} beans are instantiated
 * directly by this strategy, even if there is already an instance in the CDI context.
 * This means singletons are not really singletons, but this seems to be the behavior required by
 * the JPA 2.2 spec.
 ```

It means bean in context are not reused, it will create two instances, I don't like this behavior.

File: hibernate-core/src/main/java/org/hibernate/engine/jdbc/connections/internal/MultiTenantConnectionProviderInitiator.java (hunk 4 above)
Comment (quaff) [reply hibernate-core/src/main/java/org/hibernate/engine/jdbc/connections/internal/MultiTenantConnectionProviderInitiator.java:48]: In practice most of applications are using `DatasourceConnectionProviderImpl`, the underlying `DataSource` is managed bean, I don't think we should add complexity here.

**Comment by gavinking** [comment]
> bean in context are not reused

Because they would not be beans. That's the point.

> it will create two instances, it will create two instances

It will create one instance per persistence unit, which is _completely_ fine.

**Comment by gavinking** [comment]
>  it will create two instances

You need _what_ to be a singleton?

**Comment by quaff** [comment]
> > it will create two instances
> 
> You need _what_ to be a singleton?

The proposed `CurrentTenantIdentifierResolver` and `MultiTenantConnectionProvider`.

**Comment by yrodiere** [comment]
> > > it will create two instances
> > 
> > 
> > You need _what_ to be a singleton?
> 
> The proposed `CurrentTenantIdentifierResolver` and `MultiTenantConnectionProvider`.

This is what we were discussing yesterday, Gavin. Most people will just expect a CDI integration to involve actual CDI beans and to comply with scopes, and frankly I agree with them: it's just KISS. I know you disagree, at least in general, and I know there are technical reasons to not do that for e.g. ID generators and connection providers.

That being said, putting aside other component types, I thought you agreed it was reasonable to do that for the tenant identifier resolver?

**Comment by gavinking** [comment]
So as I was saying to Yoann on Zulip, I can see that `CurrentTenantIdentifierResolver` might legitimately be a contextual bean.

I absolutely struggle to see how this could ever possibly be legit for any sort of connection provider. Connection providers have persistence-unit-specific configuration.

And any contextual state which they might need can be injected *into* them.

**Comment by quaff** [comment]
> as that class can implement `Configurable` and get injected with PU-specific config.

OK, I get the point, I changed `useJpaCompliantCreation()` to return `true`, please review.

File: hibernate-core/src/main/java/org/hibernate/engine/jdbc/connections/internal/MultiTenantConnectionProviderInitiator.java (hunk 4 above)
Comment (quaff) [reply hibernate-core/src/main/java/org/hibernate/engine/jdbc/connections/internal/MultiTenantConnectionProviderInitiator.java:46]: I'm inclined to keep it as is for `ConnectionProviderInitiator`, WDYT @gavinking

**Comment by gavinking** [comment]
> Most people will just expect a CDI integration to involve actual CDI beans and to comply with scopes

I mean, just I don't see that.

- Servlets support CDI injection. Do you expect to be able to inject a servlet into something else? 
- Interceptors support CDI injection. Do you expect to be able to inject an interceptor into something else? 
- etc.

I don't think you _do_ expect that, so why would you expect it here?

From where would the assumption that one can inject a `MultiTenantConnectionProvider` into an arbitrary CDI bean arise? It's defined as a Hibernate `Service`, not as a CDI bean!

> I thought you agreed it was reasonable to do that for the tenant identifier resolver?

Right, correct, I think there's a reasonable argument that `CurrentTenantIdentifierResolver` is a special case here, since:

- it's _much_ more likely to be dealing with contextual state (associated with a request or session),
- it's not a Hibernate `Service`,
- it's not likely to depend on any Hibernate `Service`s or other PU-specific stats, and
- it doesn't have any defined instantiation/configuration semantics, and so Hibernate doesn't need to be involved in instantiating it.

Therefore, by making `CurrentTenantIdentifierResolver` we're providing some convenience, without tying together the lifecycle of the bean container with the instantiation of Hibernate.

I'm _much_ more concerned about `MultiTenantConnectionProvider` here.

File: hibernate-core/src/main/java/org/hibernate/engine/jdbc/connections/internal/MultiTenantConnectionProviderInitiator.java (hunk 4 above)
Comment (gavinking) [reply hibernate-core/src/main/java/org/hibernate/engine/jdbc/connections/internal/MultiTenantConnectionProviderInitiator.java:46]: I guess naively I would have said it makes sense to treat `ConnectionProvider`s and `MultiTenantConnectionProvider`s uniformly, but I'm not sure.

**Review by gavinking — APPROVED** [review]

**Comment by gavinking** [comment]
> I think there's a reasonable argument that `CurrentTenantIdentifierResolver` is a special case here

Indeed, you could choose to view the `CurrentTenantIdentifierResolver` as precisely a kind of "bridge" between the world where contextual information about the request/session is available, and the world of PU-specific state held by `Service`s.

And that explains why it's a special case.

**Comment by yrodiere** [comment]
> I mean, just I don't see that.
> 
>     * Servlets support CDI injection. Do you expect to be able to inject a servlet into something else?
> 
>     * Interceptors support CDI injection. Do you expect to be able to inject an interceptor into something else?
> 
>     * etc.
> 
> 
> I don't think you _do_ expect that, so why would you expect it here?

I do not expect to need to inject servlets, interceptors, or most pluggable Hibernate components into my application -- with the exception of tenant identifier resolvers and some others.

I do however expect that if servlets, interceptors, or pluggable Hibernate components are involved in CDI, then:

1. Trying to inject them into my application will either work, using a CDI scope as intuitive as possible, or somehow notify me (exception?) that such component just cannot be a CDI bean.
2. Putting `@ApplicationScoped` or `@RequestScoped` on them, however nonsensical, will either be taken into account, or somehow notify me (exception?) that it's invalid.

These expectations may not match current implementations, or even the specs, but I think they are reasonable from an UX point of view, and they do comply with the principle of least surprise.
They may not be achievable completely, but I think we should at least try where possible.

File: hibernate-core/src/main/java/org/hibernate/engine/jdbc/connections/internal/MultiTenantConnectionProviderInitiator.java (hunk 4 above)
Comment (quaff) [reply hibernate-core/src/main/java/org/hibernate/engine/jdbc/connections/internal/MultiTenantConnectionProviderInitiator.java:46]: I appended a commit, please review. @gavinking @beikov

File: hibernate-core/src/main/java/org/hibernate/boot/internal/SessionFactoryOptionsBuilder.java (hunk 1 above)
Comment (beikov) [reply hibernate-core/src/main/java/org/hibernate/boot/internal/SessionFactoryOptionsBuilder.java:377]: In todays Hibernate ORM team meeting, we decide to hold off on requesting actual beans from the container until JPA 4 when the question of CDI bean manager usability is settled.

File: hibernate-core/src/main/java/org/hibernate/engine/jdbc/connections/internal/MultiTenantConnectionProviderInitiator.java
Comment (beikov) [reply hibernate-core/src/main/java/org/hibernate/engine/jdbc/connections/internal/MultiTenantConnectionProviderInitiator.java:1]: In todays Hibernate ORM team meeting, we decide to hold off on requesting actual beans from the container until JPA 4 when the question of CDI bean manager usability is settled.

**Review by beikov — APPROVED** [review]

File: hibernate-core/src/main/java/org/hibernate/engine/jdbc/connections/internal/ConnectionProviderInitiator.java [hunk 5]
```diff
@@ -50,6 +56,7 @@
  * @author Steve Ebersole
  * @author Brett Meyer
  */
+@SuppressWarnings( {"deprecation", "unchecked"} )
```
Comment (gavinking) [inline hibernate-core/src/main/java/org/hibernate/engine/jdbc/connections/internal/ConnectionProviderInitiator.java:59]: Please don't suppress warnings on a whole class like this.

**Review by gavinking — CHANGES_REQUESTED** [review]

File: hibernate-core/src/main/java/org/hibernate/engine/jdbc/connections/internal/ConnectionProviderInitiator.java (hunk 5 above)
Comment (quaff) [reply hibernate-core/src/main/java/org/hibernate/engine/jdbc/connections/internal/ConnectionProviderInitiator.java]: done.

File: hibernate-core/src/main/java/org/hibernate/boot/internal/SessionFactoryOptionsBuilder.java [hunk 6]
```diff
@@ -372,6 +377,38 @@ public SessionFactoryOptionsBuilder(StandardServiceRegistry serviceRegistry, Boo
...
+							}
+
+							@Override
+							public <B> B produceBeanInstance(String name, Class<B> beanType) {
+								return null;
+							}
+
+						}
+				).getBeanInstance();
+			}
```
Comment (gavinking) [inline hibernate-core/src/main/java/org/hibernate/boot/internal/SessionFactoryOptionsBuilder.java:410]: Please extract out a function for this, the constructor is already waaaaaay too long.

File: hibernate-core/src/main/java/org/hibernate/engine/jdbc/connections/internal/ConnectionProviderInitiator.java [hunk 7]
```diff
@@ -177,11 +184,47 @@ else if ( configurationValues.containsKey( URL ) ) {
...
+							}
+
+						}
+				).getBeanInstance();
+			}
+			else {
+				return noAppropriateConnectionProvider();
+			}
+
 		}
```
Comment (gavinking) [inline hibernate-core/src/main/java/org/hibernate/engine/jdbc/connections/internal/ConnectionProviderInitiator.java:220]: Similarly, please extract a function.

**Review by gavinking — CHANGES_REQUESTED** [review]
@quaff, thanks, but please see my further comments.

File: hibernate-core/src/main/java/org/hibernate/boot/internal/SessionFactoryOptionsBuilder.java (hunk 6 above)
Comment (quaff) [reply hibernate-core/src/main/java/org/hibernate/boot/internal/SessionFactoryOptionsBuilder.java:410]: @gavinking I suggest to introduce new util methods for `org.hibernate.resource.beans.internal.Helper` and refactor all `BeanContainer` access to use those methods in another PR, WDYT?

```java
	@SuppressWarnings( "unchecked" )
	public static <T> T getBean(ServiceRegistry serviceRegistry, Class<?> beanType, boolean canUseCachedReferences, boolean useJpaCompliantCreation, T fallback) {
		final BeanContainer beanContainer = allowExtensionsInCdi( serviceRegistry ) ? serviceRegistry.requireService( ManagedBeanRegistry.class ).getBeanContainer() : null;
		if ( beanContainer == null ) {
			return null;
		}
		return (T) beanContainer.getBean(
				beanType,
				new BeanContainer.LifecycleOptions() {
					@Override
					public boolean canUseCachedReferences() {
						return canUseCachedReferences;
					}

					@Override
					public boolean useJpaCompliantCreation() {
						return useJpaCompliantCreation;
					}
				},
				new BeanInstanceProducer() {

					@Override
					public <B> B produceBeanInstance(Class<B> beanType) {
						return (B) fallback;
					}

					@Override
					public <B> B produceBeanInstance(String name, Class<B> beanType) {
						throw new UnsupportedOperationException("The method shouldn't be called");
					}

				}
		).getBeanInstance();
	}
```

Comment (gavinking) [reply hibernate-core/src/main/java/org/hibernate/boot/internal/SessionFactoryOptionsBuilder.java:410]: >  I suggest to introduce new util methods for org.hibernate.resource.beans.internal.Helper and refactor all `BeanContainer` access to use those methods

That sounds fine to me, but please do it now, so that it doesn't get forgotten. Thanks.

Comment (quaff) [reply hibernate-core/src/main/java/org/hibernate/boot/internal/SessionFactoryOptionsBuilder.java:410]: I created https://hibernate.atlassian.net/browse/HHH-18737 just now.

Comment (gavinking) [reply hibernate-core/src/main/java/org/hibernate/boot/internal/SessionFactoryOptionsBuilder.java:410]: OK, great, thanks 🙏
